src/Symfony/Component/ExpressionLanguage/Lexer.php

```diff
@@ -40,11 +40,11 @@ public function tokenize(string $expression)
                 continue;
             }
 
-            if (preg_match('/[0-9]+(?:\.[0-9]+)?([Ee][\+\-][0-9]+)?/A', $expression, $match, 0, $cursor)) {
+            if (preg_match('/((?:[0-9]+(?:_[0-9]+)*|(?:(?:[0-9]+(?:_[0-9]+)*)?\.[0-9]+(?:_[0-9]+)*|[0-9]+(?:_[0-9]+)*\.(?!\.)(?:[0-9]+(?:_[0-9]+)*)?))[eE][+-]?[0-9]+(?:_[0-9]+)*|(?:(?:[0-9]+(?:_[0-9]+)*)?\.[0-9]+(?:_[0-9]+)*|[0-9]+(?:_[0-9]+)*\.(?!\.)(?:[0-9]+(?:_[0-9]+)*)?)|[0-9]+(?:_[0-9]+)*)/A', $expression, $match, 0, $cursor)) {
```

Comment:
I had to add to a negative lookahead (`(?!\.)`) because the base pattern from the PHP source code matches decimals finishing at the point (eg: `8.`) and it collides with the range operator (eg: `0..9` would be interpreted as number `0` + number `0.9` instead of number `0` + operator `..` + number `9`)

Reply:
:vomiting_face: :smile: 
Can't we simplify this by not implementing support for lexing decimals with no leading zero?
If yes, I'd be fine closing this PR (keeping it for the record) and adding support for `_` in another PR.
What do others think?

Reply:
If we keep that:
`(?(DEFINE)(?P<LNUM>[0-9]+(_[0-9]+)*))`
Then `(?&LNUM)` instead of the corresponding pattern?